codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java

```diff
@@ -211,19 +366,108 @@ int streamableBytes0(Http2Stream stream) {
+        State parent = state(parentId);
+        State child;
+        return parent.children.containsKey(childId) &&
+                (child = state(childId)).parent == parent && child.weight == weight;
+    }
+
+    /**
+     * For testing only!
+     */
+    int numChildren(int streamId) {
+        State state = state(streamId);
+        return state == null ? 0 : state.children.size();
+    }
+
+    /**
+     * Notify all listeners of the priority tree change events (in ascending order)
+     * @param events The events (top down order) which have changed
+     */
+    void notifyParentChanged(List<ParentChangedEvent> events) {
+        for (ParentChangedEvent event : events) {
+            stateOnlyRemovalQueue.priorityChanged(event.state);
+            if (event.state.parent != null && event.state.activeCountForTree != 0) {
+                event.state.parent.offerAndInitializePseudoTime(event.state);
+                event.state.parent.activeCountChangeForTree(event.state.activeCountForTree);
+            }
+        }
+    }
+
+    /**
+     * A comparator for {@link State} which has no associated {@link Http2Stream} object. The general precedence is:
+     * <ul>
+     *     <li>Was a stream activated or reserved (streams only used for priority are higher priority)</li>
+     *     <li>Depth in the priority tree (closer to root is higher priority></li>
+     *     <li>Stream ID (higher stream ID is higher priority - used for tie breaker)</li>
+     * </ul>
+     */
+    private static final class StateOnlyComparator implements Comparator<State> {
+        static final StateOnlyComparator INSTANCE = new StateOnlyComparator();
+
+        private StateOnlyComparator() {
+        }
+
+        @Override
+        public int compare(State o1, State o2) {
+            // "priority only streams" (which have not been activated) are higher priority than streams used for data.
+            boolean o1Actived = o1.wasStreamReservedOrActivated();
+            if (o1Actived != o2.wasStreamReservedOrActivated()) {
+                return o1Actived ? -1 : 1;
+            }
+            // Numerically greater depth is higher priority.
+            int x = o2.dependencyTreeDepth - o1.dependencyTreeDepth;
+
+            // I also considered tracking the number of streams which are "activated" (eligible transfer data) at each
+            // subtree. This would require a traversal from each node to the root on dependency tree structural changes,
+            // and then it would require a re-prioritization at each of these nodes (instead of just the nodes where the
+            // direct parent changed). The costs of this are judged to be relatively high compared to the nominal
+            // benefit it provides to the heuristic. Instead folks should just increase maxStateOnlySize.
+
+            // Last resort is to give larger stream ids more priority.
+            return x != 0 ? x : o1.streamId - o2.streamId;
```

Comment:
The values are constrained to be positive so there shouldn't be any overflow and subtraction is preferred over a method call with multiple conditional statements.

Also `Integer.compare` is JDK7+ so we would need to replicate this in `MathUtil` to retain JDK6+ compatibility. I debated doing this but I think the current approach is better.